Description: Fix wrong homedir permissions when UMASK is unset in /etc/login.defs
 When a user doesn't exist and user module is used to create the user and the
 homedir, adduser is called which parses HOME_MODE from /etc/login.defs, and
 when not set calculates the mode from UMASK from the same file.
 .
 When a user already exists without homedir, and the user module is used to add
 a home dir, it incorrectly ignores HOME_MODE, resulting in a world-readable
 home dir when UMASK is not set. This is for example the case in Debian trixie
 and later, and likely Ubuntu 25.04 and later.
 .
 Finally, when neither HOME_MODE nor UMASK was set, this module threw an
 exception. So instead, follow adduser behaviour and initialize UMASK = 022.
Author: Lee Garrett <debian@rocketjump.eu>
Origin: upstream, https://patch-diff.githubusercontent.com/raw/ansible/ansible/pull/84482.patch
Forwarded: https://github.com/ansible/ansible/pull/84482
Last-Update: 2024-12-20
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/

diff --git a/lib/ansible/modules/user.py b/lib/ansible/modules/user.py
index 8fdc71aae8ce82..2d4b9ecf8f31b6 100644
--- a/lib/ansible/modules/user.py
+++ b/lib/ansible/modules/user.py
@@ -1372,14 +1372,25 @@ def create_homedir(self, path):
             if os.path.exists(self.LOGIN_DEFS):
                 with open(self.LOGIN_DEFS, 'r') as f:
                     for line in f:
+                        # HOME_MODE has higher precedence as UMASK
+                        m = re.match(r'^HOME_MODE\s+(\d+)$', line)
+                        if m:
+                            mode = int(m.group(1), 8)
+                            break  # higher precedence
                         m = re.match(r'^UMASK\s+(\d+)$', line)
                         if m:
                             umask = int(m.group(1), 8)
                             mode = 0o777 & ~umask
-                            try:
-                                os.chmod(path, mode)
-                            except OSError as e:
-                                self.module.exit_json(failed=True, msg="%s" % to_native(e))
+                # fallback if neither HOME_MODE nor UMASK are set;
+                # follow behaviour of useradd initializing UMASK = 022
+                try:
+                    mode
+                except NameError:
+                    mode = 0o755
+                try:
+                    os.chmod(path, mode)
+                except OSError as e:
+                    self.module.exit_json(failed=True, msg="%s" % to_native(e))
 
     def chown_homedir(self, uid, gid, path):
         try:
diff --git a/test/integration/targets/user/tasks/test_no_home_fallback.yml b/test/integration/targets/user/tasks/test_no_home_fallback.yml
index f7627fae1e39b0..0783ec1b6c8faf 100644
--- a/test/integration/targets/user/tasks/test_no_home_fallback.yml
+++ b/test/integration/targets/user/tasks/test_no_home_fallback.yml
@@ -35,12 +35,17 @@
         import os
         try:
             for line in open('/etc/login.defs').readlines():
+                m = re.match(r'^HOME_MODE\s+(\d+)$', line)
+                if m:
+                    mode = oct(int(m.group(1), 8))
+                    break
                 m = re.match(r'^UMASK\s+(\d+)$', line)
                 if m:
                     umask = int(m.group(1), 8)
+                    mode = oct(0o777 & ~umask)
         except:
             umask = os.umask(0)
-        mode = oct(0o777 & ~umask)
+            mode = oct(0o777 & ~umask)
         print(str(mode).replace('o', ''))
       args:
         executable: "{{ ansible_python_interpreter }}"
